### PR TITLE
ESQL: Method to convert BooleanBlock to a "mask"

### PR DESCRIPTION
**Pull request by nik9000**
This adds a method, `BooleanBlock#toMask` to convert `BooleanBlock`s into a "mask" for use with `keepMask`.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayBlockBuilderTests.java [hunk 1]
```diff
@@ -266,4 +284,48 @@ public void testBooleanBlock() throws IOException {
...
+                    assertThat(mask.mask().getBoolean(0), equalTo(false));
+                    for (int p = 1; p < positionCount; p++) {
+                        assertThat(mask.mask().getBoolean(p), equalTo(elements[mvCount + p]));
+                    }
+                }
+            }
+        }
+        assertThat(blockFactory.breaker().getUsed(), equalTo(0L));
+    }
 }
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayBlockBuilderTests.java:331]: nit: there's 2 ways of having null values in an array block:
- the nullsmask is explicitly set at the position
- the value count is just 0.

I'm not sure if the added tests cover `toMask` in the second case.

**Review by alex-spies — APPROVED** [review]
LGTM, only minor remarks!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-ArrayBlock.java.st [hunk 2]
```diff
@@ -101,6 +101,32 @@ $if(BytesRef)$
     public OrdinalBytesRefBlock asOrdinals() {
         return null;
     }
+
+$elseif(boolean)$
+    /**
+     * Convert this to a {@link BooleanVector "mask"} that's appropriate for
+     * passing to {@link #keepMask}.
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-ArrayBlock.java.st:108]: :+1:

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/ToMask.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+package org.elasticsearch.compute.data;
+
+import org.elasticsearch.core.Releasable;
+
+/**
+ * Result from calling {@link BooleanBlock#toMask}. {@link #close closing} this will
+ * close the contained {@link #mask()}. If you want to keep a reference to it then you'll
+ * have to {@link Block#incRef()} it.
+ */
+public record ToMask(BooleanVector mask, boolean hadMultivaluedFields) implements Releasable {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/ToMask.java:17]: It contains the second return value - the `boolean` for whether the thing had multivalued fields that were squished away. So we can warn about it.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BigArrayVector.java.st [hunk 4]
```diff
@@ -84,6 +84,18 @@ $endif$
         return new $Type$VectorBlock(this);
     }
 
+$if(boolean)$
+    /**
+     * Convert this to a {@link BooleanVector "mask"} that's appropriate for
+     * passing to {@link #keepMask}.
+     */
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BigArrayVector.java.st]: I just pushed that bit.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayBlockBuilderTests.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayBlockBuilderTests.java:335]: The test never makes 0 value count. Also! Those *should* mean the same thing.

**Review by ChrisHegarty — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayBlockBuilderTests.java (hunk 1 above)
Comment (alex-spies) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayBlockBuilderTests.java:335]: That's my point though: for good measure, I think the test should make a block with a 0 value count at `p` (probably without setting the nulls mask at `p`); `toMask` should still work. (I'm pretty sure it does rn, but tests are better proof than my word!)

Last time I went digging in these parts of the code (when I added the invariant check for array blocks), there were rare instances where we actually made 0 value counts (by setting the first index for `p` and `p+1` to the same value). Not sure if that's still the case, but I think we at least allow it.

Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayBlockBuilderTests.java:335]: I'll open an PR. I think at this point not setting the null mask but setting the value count is an error. All kinds of things will blow up.

**Pull request merged**